Signal/src/ViewControllers/ConversationView/Cells/OWSExpirationTimerView.m

```diff
@@ -4,205 +4,179 @@
-    self.emptyHourglassImageView.frame = hourglassFrame;
-    self.emptyHourglassImageView.bounds = hourglassFrame;
-    self.fullHourglassImageView.frame = hourglassFrame;
-    self.fullHourglassImageView.bounds = hourglassFrame;
+    [self.layer removeAllAnimations];
+    [self.maskLayer removeAllAnimations];
+    [self.maskLayer removeFromSuperlayer];
+    self.maskLayer = nil;
+    [self.fullHourglassImageView.layer.mask removeFromSuperlayer];
+    self.fullHourglassImageView.layer.mask = nil;
+    self.layer.opacity = 1.f;
+    self.emptyHourglassImageView.hidden = YES;
+    self.fullHourglassImageView.hidden = YES;
+    [self.animationTimer invalidate];
+    self.animationTimer = nil;
 }
 
-- (void)handleReappearNotification:(NSNotification *)notification
-{
-    DDLogVerbose(@"%@ handleReappearNotification", self.logTag);
-    [self startAnimation];
+- (void)setFrame:(CGRect)frame {
+    BOOL sizeDidChange = CGSizeEqualToSize(self.frame.size, frame.size);
+    [super setFrame:frame];
+    if (sizeDidChange) {
+        [self ensureAnimations];
+    }
 }
 
-- (void)startTimerWithExpiresAtSeconds:(double)expiresAtSeconds initialDurationSeconds:(uint32_t)initialDurationSeconds
-{
-    if (expiresAtSeconds == 0) {
-        DDLogWarn(
-            @"%@ Asked to animate expiration for message which hasn't started expiring. intitialDurationSeconds:%u",
-            self.logTag,
-            initialDurationSeconds);
+- (void)setBounds:(CGRect)bounds {
+    BOOL sizeDidChange = CGSizeEqualToSize(self.bounds.size, bounds.size);
+    [super setBounds:bounds];
+    if (sizeDidChange) {
+        [self ensureAnimations];
     }
-
-    DDLogVerbose(@"%@ Starting timer with expiresAtSeconds: %f initialDurationSeconds: %d",
-        self.logTag,
-        expiresAtSeconds,
-        initialDurationSeconds);
-
-    self.expiresAtSeconds = expiresAtSeconds;
-    self.initialDurationSeconds = initialDurationSeconds;
-    [[NSNotificationCenter defaultCenter] addObserver:self
-                                             selector:@selector(handleReappearNotification:)
-                                                 name:ConversationViewControllerDidAppearNotification
-                                               object:nil];
-    [self startAnimation];
 }
 
-- (void)startAnimation
+- (void)ensureAnimations
 {
-    DDLogVerbose(@"%@ Starting animation with expiresAtSeconds: %f initialDurationSeconds: %d",
-        self.logTag,
-        self.expiresAtSeconds,
-        self.initialDurationSeconds);
+    OWSAssert([NSThread isMainThread]);
+    
+    CGFloat secondsLeft = MAX(0, (self.expirationTimestamp - [NSDate ows_millisecondTimeStamp]) / 1000.f);
 
-    double secondsLeft = self.expiresAtSeconds - [NSDate new].timeIntervalSince1970;
+    [self clearAnimations];
+    
+    const NSTimeInterval kBlinkAnimationDurationSeconds = 2;
 
-    if (secondsLeft < 0) {
-        secondsLeft = 0;
+    if (self.expirationTimestamp == 0) {
```

Comment:
There's three cases, I think it's bit more readable this way.

Reply:
👍